leaderboard/llvm_codesize/e_greedy/e_greedy.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""ϵ-greedy policy for LLVM codesize."""
+import logging
+import os
+import random
+import sys
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import NamedTuple
+
+from absl import flags
+
+from compiler_gym.envs import CompilerEnv, LlvmEnv
+
+# Import the ../eval_policy.py helper.
+sys.path.insert(0, os.path.dirname(os.path.realpath(__file__)) + "/..")
+from eval_policy import eval_policy  # noqa pylint: disable=wrong-import-position
+
+flags.DEFINE_float(
+    "epsilon", 0, "The ratio of patience to the size of the action space. "
+)
+FLAGS = flags.FLAGS
+
+
+class ActionReward(NamedTuple):
+    """An action -> reward tuple for a single step()."""
+
+    action: int
+    reward: float
+
+
+def select_best_action(env: CompilerEnv, executor: ThreadPoolExecutor) -> ActionReward:
+    """Determine the best action by trying all possible options and ranking them."""
+
+    def eval_action(fkd: CompilerEnv, action: int) -> ActionReward:
+        """Evaluate the given action."""
+        try:
+            _, reward, _, _ = fkd.step(action)
+        finally:
+            fkd.close()
+        return ActionReward(action=action, reward=reward)
+
+    # Create an ordered list of actions, best first, ordered by descding reward,
+    # using action index as a tie-breaker. Do this by creating n forks of the
+    # environment, one for every action, and evaluting the actions in parallel
+    # threads. Note that calls to fork() occur in the main thread for thread
+    # safety in case of environment restart.
+    futures = (
```

Comment:
This is fancy and looks cool!

However I wonder if it'd be faster to implement a revert_step() for the environment to return to previous state? Maybe compared to forking another environment could be faster? 

But obviously forking is a more general option.

Reply:
Yeah a `revert_step()` would be super useful, but the problem is that the compiler passes don't have an "undo" function, so the only way to do this would be to call `env.fork()` before every step, building up a stack of environments, one for each step, and to pop from the top on `revert_step()`. This is totally do-able btw